File: 016_3sum_closest/Solution.py
class Solution:
    def threeSumClosest(self, nums: list, target: int) -> int:
        list_len = len(nums)
        if not nums or list_len < 3: return 0

        nums.sort()
        min_diff = float('inf')
        result = 0

        for i in range(list_len):
            l = i + 1
            r = list_len - 1
            while l < r:
                cur_sum = nums[i] + nums[l] + nums[r]
                if cur_sum == target: return cur_sum
                cur_diff = abs(cur_sum - target)
                if cur_diff < min_diff:
                    result = cur_sum
                    min_diff = cur_diff

                if cur_sum > target: r -= 1
                else: l += 1

        return result


if __name__ == '__main__':
    solution = Solution()
    nums = [0, 1, 2]
    target = 0
    print(solution.threeSumClosest(nums, target))

File: 016_3sum_closest/Solution.go
package main

import (
    "fmt"
    "math"
    "sort"
)

func threeSumClosest(nums []int, target int) int {
    arrLen := len(nums)
    if nums == nil || arrLen < 3 {
        return 0
    }

    sort.Ints(nums)
    result, curSum, curDiff := 0, 0, 0
    minDiff := math.MaxInt32

    for i, v := range nums {
        L, R := i+1, arrLen-1
        for L < R {
            curSum = v + nums[L] + nums[R]
            if curSum == target {
                return curSum
            }
            curDiff = int(math.Abs(float64(curSum - target)))
            if curDiff < minDiff {
                result = curSum
                minDiff = curDiff
            }

            if curSum > target {
                R--
            } else {
                L++
            }
        }
    }

    return result
}

func main() {
    nums := []int{-1, 0, 1, 1, 55}
    target := 3
    fmt.Println(threeSumClosest(nums, target))
}

